Refactor plugin structure

Valemos opened the pull request:
Why:
Remove unused code clutter to avoid possible race conditions. Structure plugin to be easily maintainable. 
Tested with real Bambu printer to ensure correct interaction.
Changes:
- Add separate class for sftp file system
- Add separate serial IO handling class
- Replace function name mangling with gcode handler registration system
- Add states to virtual Bambu printer that manage state specific interaction
- Add synchronization utilities to work with virtual printer as if it is a binary stream
- Add unittests with mocked Bambu printer to ensure core functionality works as expected
- Fix formatting to be automatically processed by black formatter
- Fix python 3.10 type annotations for readability

Valemos commented:
I wanted to refactor this plugin, since I want to add gcode view support.
I did this refactor as separate pull request to keep these operations separate.

Valemos commented:
I'm almost done. I want to move most bambu printer specific logic from plugin class to printer class (timelapses and file upload/delete)
You can start reviewing all classes except RemoteSDCardFileList and BambuPrintPlugin

jneilliii commented:
Really good job on the refactor. I've tested it with my X1C and nothing stands out as being problematic. I am seeing one weird thing with logging, seems to be duplication in extremely short periods of time. Haven't dug too deep yet, but I assume it's due to logging in different parts of the refactor. 

```
2024-07-24 23:36:47,760 - octoprint.plugins.bambu_printer.BambuPrinter.serial - DEBUG - ------------------------------------------------------------------------------
2024-07-24 23:36:47,771 - octoprint.plugins.bambu_printer.BambuPrinter - DEBUG - connecting via local mqtt: True
2024-07-24 23:36:47,776 - octoprint.plugins.bambu_printer.BambuPrinter - DEBUG - connecting via local mqtt: True
2024-07-24 23:36:47,787 - octoprint.plugins.bambu_printer.BambuPrinter - DEBUG - on disconnect called
2024-07-24 23:36:47,793 - octoprint.plugins.bambu_printer.BambuPrinter - DEBUG - on disconnect called
2024-07-24 23:36:47,794 - octoprint.plugins.bambu_printer.BambuPrinter - DEBUG - on connect called
2024-07-24 23:36:47,797 - octoprint.plugins.bambu_printer.BambuPrinter - DEBUG - on connect called
2024-07-24 23:36:47,894 - octoprint.plugins.bambu_printer.BambuPrinter - INFO - bambu connection status: False
2024-07-24 23:36:47,987 - octoprint.plugins.bambu_printer.BambuPrinter - INFO - bambu connection status: False
```

Valemos commented:
@jneilliii 
Done with ftps file system. I have not checked the timelapse logic with real hardware. Could you check on your side?
Also, maybe it makes sense to enable bambu camera preview by default? Not aware how this functionality can be added correctly into plugin. Maybe you can elaborate to make it faster?

jneilliii commented:
> Also, maybe it makes sense to enable bambu camera preview by default? Not aware how this functionality can be added correctly into plugin. Maybe you can elaborate to make it faster?

This is actually not that straight forward based on the device. For example, for the X1 I use go2rtc to transcode rtsp feed into a stream that works with classic webcam plugin (using HLS). However, the A1/P1 series devices rtsp feed are kind of crappy from what I've been told. It's still possible but very low fps. @synman figured it out though using go2rtc exec stream and access may be blocked locally if in cloud mode, etc. You can find more on his repo [here](https://github.com/synman/bambu-go2rtc) related to those printers.

jneilliii commented:
> I have not checked the timelapse logic with real hardware. Could you check on your side?

This is definitely broken. I haven't dug too deep into it yet to know why, but the files aren't getting listed on the timelapse tab in the Bambu Timelapses section like it was before.

jneilliii commented:
I also get this error on initial startup of OctoPrint, not sure if it's related. The SD card file listing of 3mf files seem to be working fine.

```
Traceback (most recent call last):
  File "C:\Program Files\Python39\lib\threading.py", line 917, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\jneil\Documents\3D Prints\PrinterUprades\OrdBotHadron\OctoprintRelated\OctoPrint-BambuPrinter\octoprint_bambu_printer\bambu_print_plugin.py", line 199, in process
    for file_info in self._bambu_file_system.get_all_timelapse_files():
  File "C:\Users\jneil\Documents\3D Prints\PrinterUprades\OrdBotHadron\OctoprintRelated\OctoPrint-BambuPrinter\octoprint_bambu_printer\printer\file_system\remote_sd_card_file_list.py", line 55, in get_all_timelapse_files
    self._timelapse_files_view.update()
  File "C:\Users\jneil\Documents\3D Prints\PrinterUprades\OrdBotHadron\OctoprintRelated\OctoPrint-BambuPrinter\octoprint_bambu_printer\printer\file_system\cached_file_view.py", line 40, in update
    file_info_list = self.list_all_views()
  File "C:\Users\jneil\Documents\3D Prints\PrinterUprades\OrdBotHadron\OctoprintRelated\OctoPrint-BambuPrinter\octoprint_bambu_printer\printer\file_system\cached_file_view.py", line 36, in list_all_views
    result.extend(self.file_system.list_files(*filter, ftp, existing_files))
  File "C:\Users\jneil\Documents\3D Prints\PrinterUprades\OrdBotHadron\OctoprintRelated\OctoPrint-BambuPrinter\octoprint_bambu_printer\printer\file_system\remote_sd_card_file_list.py", line 93, in list_files
    return list(
  File "C:\Users\jneil\Documents\3D Prints\PrinterUprades\OrdBotHadron\OctoprintRelated\OctoPrint-BambuPrinter\octoprint_bambu_printer\printer\file_system\remote_sd_card_file_list.py", line 131, in get_file_info_for_names
    file_info = self._get_ftp_file_info(ftp, entry, existing_files)
  File "C:\Users\jneil\Documents\3D Prints\PrinterUprades\OrdBotHadron\OctoprintRelated\OctoPrint-BambuPrinter\octoprint_bambu_printer\printer\file_system\remote_sd_card_file_list.py", line 105, in _get_ftp_file_info
    file_size = ftp.ftps_session.size(file_path.as_posix())
  File "C:\Program Files\Python39\lib\ftplib.py", line 630, in size
    resp = self.sendcmd('SIZE ' + filename)
  File "C:\Program Files\Python39\lib\ftplib.py", line 281, in sendcmd
    return self.getresp()
  File "C:\Program Files\Python39\lib\ftplib.py", line 254, in getresp
    raise error_perm(resp)
ftplib.error_perm: 550 Could not get file size.
```

jneilliii commented:
I think maybe the path to the file request for timelapse folder is getting duplicated somehwere....

![image](https://github.com/user-attachments/assets/bab10718-f023-4b26-ac42-008d50b250c2)

jneilliii commented on octoprint_bambu_printer/printer/file_system/ftps_client.py:
```diff
@@ -185,19 +144,26 @@ def move_file(self, source: str, dest: str):
-            return files
+            list_result = self.ftps_session.nlst(list_path) or []
+            for file_name in list_result:
+                path = Path(list_path) / file_name
+                if _extension_acceptable(path):
+                    yield path
```
I think this is where the issue is with the timelapse folder listing stuff and the duplicated pathing.

jneilliii replied:
In the original code I had some stupid logic to get around this. I can't remember if it's because of a difference in how X1 list files on ftps vs A1/P1 series printers.

https://github.com/jneilliii/OctoPrint-BambuPrinter/blob/ad862d5ebd4b0579d52923c2725ea4c7a7cc137e/octoprint_bambu_printer/__init__.py#L154-L158

Valemos replied:
The equivalent of this operation would be to use Path(filename).name this accounts for most operations you've done

Valemos replied:
I see the issue with that filename. I use parent folder as base and append a filename to it. but filename already includes folder name, since listing files for p1s gives only names, but in your case maybe it lists them with folder name

jneilliii replied:
Exactly this, the different printer models give the list in different ways. Maybe the more generic solution to my hack is replacing list_path in filename...

```
path = Path(list_path) / file_name.replace(list_path, "")
```

Valemos replied:
Fixed this issue, covered with tests and improved logging for ftp connection.

jneilliii replied:
yep, works perfect now with timelapse listing and loading thumbnails. running through some real world tests now with a print.

jneilliii commented:
started print from orca slicer and it was picked up as printing, but there are repeating file opened/selected lines over and over. although it might not pose an issue, these should only be sent at the initial detection of printing (started externally) or when starting a print. The only thing that should be consistently reported is the `SD printing byte <current byte>/<total bytes>`.

```
Recv: File opened: Fliegenpistole Remix v17.gcode.3mf  Size: 298154
Recv: File selected
Recv: SD printing byte 35778/298154
Send: N439 M27*24
Recv: SD printing byte 35778/298154
Recv: ok
Send: N440 M105*39
Recv: T:219.00/ 220.00 B:35.00/ 35.00 @:64
Recv: ok
Send: N441 M27*23
Recv: SD printing byte 35778/298154
Recv: ok
Send: N442 M27*20
Recv: SD printing byte 35778/298154
Recv: ok
Recv: File opened: Fliegenpistole Remix v17.gcode.3mf  Size: 298154
Recv: File selected
Recv: SD printing byte 35778/298154
Send: N443 M27*21
Recv: SD printing byte 35778/298154
Recv: ok
Send: N444 M27*18
Recv: SD printing byte 35778/298154
Recv: ok
Send: N445 M27*19
Recv: SD printing byte 35778/298154
Recv: ok
Recv: File opened: Fliegenpistole Remix v17.gcode.3mf  Size: 298154
Recv: File selected
Recv: SD printing byte 35778/298154
Send: N446 M105*33
Recv: T:220.00/ 220.00 B:35.00/ 35.00 @:64
Recv: ok
Send: N447 M27*17
Recv: SD printing byte 35778/298154
Recv: ok
Send: N448 M27*30
Recv: SD printing byte 35778/298154
Recv: ok
Send: N449 M27*31
Recv: SD printing byte 35778/298154
Recv: ok
Recv: File opened: Fliegenpistole Remix v17.gcode.3mf  Size: 298154
Recv: File selected
Recv: SD printing byte 35778/298154
Send: N450 M27*23
Recv: SD printing byte 35778/298154
Recv: ok
Send: N451 M27*22
Recv: SD printing byte 35778/298154
Recv: ok
Send: N452 M105*36
Recv: T:220.00/ 220.00 B:35.00/ 35.00 @:64
Recv: ok
Send: N453 M27*20
Recv: SD printing byte 35778/298154
Recv: ok
Recv: File opened: Fliegenpistole Remix v17.gcode.3mf  Size: 298154
Recv: File selected
Recv: SD printing byte 38760/298154
Send: N454 M27*19
Recv: SD printing byte 38760/298154
Recv: ok
Send: N455 M27*18
Recv: SD printing byte 38760/298154
Recv: ok
Send: N456 M27*17
Recv: SD printing byte 38760/298154
Recv: ok
Recv: File opened: Fliegenpistole Remix v17.gcode.3mf  Size: 298154
Recv: File selected
Recv: SD printing byte 38760/298154
Send: N457 M27*16
Recv: SD printing byte 38760/298154
Recv: ok
Send: N458 M105*46
Recv: T:220.00/ 220.00 B:35.00/ 35.00 @:64
Recv: ok
Send: N459 M27*30
Recv: SD printing byte 38760/298154
Recv: ok
Send: N460 M27*20
Recv: SD printing byte 38760/298154
Recv: ok
Recv: File opened: Fliegenpistole Remix v17.gcode.3mf  Size: 298154
Recv: File selected
Recv: SD printing byte 38760/298154
Send: N461 M27*21
Recv: SD printing byte 38760/298154
Recv: ok
Send: N462 M27*22
Recv: SD printing byte 38760/298154
Recv: ok
Send: N463 M27*23
Recv: SD printing byte 38760/298154
Recv: ok
Recv: File opened: Fliegenpistole Remix v17.gcode.3mf  Size: 298154
Recv: File selected
Recv: SD printing byte 38760/298154
Send: N464 M105*33
Recv: T:220.00/ 220.00 B:35.00/ 35.00 @:64
Recv: ok
Send: N465 M27*17
Recv: SD printing byte 38760/298154
Recv: ok
Send: N466 M27*18
Recv: SD printing byte 38760/298154
Recv: ok
Send: N467 M27*19
Recv: SD printing byte 38760/298154
Recv: ok
Recv: File opened: Fliegenpistole Remix v17.gcode.3mf  Size: 298154
Recv: File selected
Recv: SD printing byte 38760/298154
Send: N468 M27*28
Recv: SD printing byte 38760/298154
Recv: ok
Send: N469 M27*29
Recv: SD printing byte 38760/298154
Recv: ok
```

jneilliii commented:
Pausing a print works, but resuming seems to think it worked, but it doesn't actually resume. I had to go into orca slicer to resume the print.

jneilliii commented:
Print completion is also not detected. Had a print run overnight and it's done, but OctoPrint is still showing it as printing.

![image](https://github.com/user-attachments/assets/16926e1a-8d22-4428-bc0f-e4b5e19f6dc8)

![image](https://github.com/user-attachments/assets/f57a610a-be8e-44c2-bb69-3ab911ca8383)

jneilliii commented on octoprint_bambu_printer/printer/states/paused_state.py:
```diff
@@ -0,0 +1,53 @@
+        )
+        paused_timer.start()
+
+    def resume_print(self):
```
I think changing this from resume_print to start_new_print will resolve the issue with resuming. With SD card printing the same gcode command (M24) is used for starting and resuming. So that way this function will call the resume command when in paused state, assuming I'm following the refactor logic properly. 

```
    @gcode_executor.register_no_data("M24")
    def _start_print(self):
        self._current_state.start_new_print()
        return True
```

jneilliii commented:
Attempting to start a print from OctoPrint ended up in error pop up in OrcaSlicer.

![image](https://github.com/user-attachments/assets/47cb98dc-c061-41dc-8f0a-565dea06adc0)

```
2024-07-27 15:05:59,661 - octoprint.plugins.bambu_printer.BambuPrinter - DEBUG - processing gcode M24 command = M24
2024-07-27 15:05:59,661 - octoprint.plugins.bambu_printer.BambuPrinter.gcode_executor - DEBUG - Executing M24 (Start or Resume SD print)
2024-07-27 15:05:59,672 - octoprint.plugins.bambu_printer.BambuPrinter.states - INFO - Started print for SpiceRackV2.gcode.3mf
2024-07-27 15:05:59,673 - octoprint.plugins.bambu_printer.BambuPrinter - DEBUG - Changing state from IdleState to PrintingState
2024-07-27 15:05:59,673 - octoprint.plugins.bambu_printer.BambuPrinter - DEBUG - Select project file: SpiceRackV2.gcode.3mf
2024-07-27 15:05:59,676 - octoprint.plugins.bambu_printer.BambuPrinter.states - DEBUG - SD File Print finishing: SpiceRackV2.gcode.3mf
2024-07-27 15:05:59,678 - octoprint.plugins.bambu_printer.BambuPrinter - DEBUG - Changing state from PrintingState to IdleState
2024-07-27 15:05:59,676 - octoprint.plugins.bambu_printer.BambuPrinter.serial - DEBUG - >>> ok
2024-07-27 15:05:59,681 - octoprint.util.comm - INFO - Changing monitoring state from "Starting print from SD" to "Printing from SD"
2024-07-27 15:05:59,680 - octoprint.plugins.bambu_printer.BambuPrinter.serial - DEBUG - >>> File opened: SpiceRackV2.gcode.3mf  Size: 3473171
2024-07-27 15:05:59,686 - octoprint.plugins.bambu_printer.BambuPrinter.serial - DEBUG - >>> File selected
2024-07-27 15:05:59,695 - octoprint.printer.standard.job - INFO - Print job selected - origin: sdcard, path: SpiceRackV2.gcode.3mf, owner: None, user: None
2024-07-27 15:05:59,700 - octoprint.plugins.bambu_printer.BambuPrinter.serial - DEBUG - >>> Done printing file
```

jneilliii commented:
> Attempting to start a print from OctoPrint ended up in error pop up in OrcaSlicer.

I think this is a conversion issue for print_command properties...

![image](https://github.com/user-attachments/assets/932a3d7d-420b-421f-aa23-e919b1821f52)

jneilliii commented:
I think the above commit fixes the start print issue.

jneilliii commented:
noticed a cancelled print continues to report sd printing in Operational state, usually this switches to "not sd printing".

![image](https://github.com/user-attachments/assets/4039c72e-0d45-4d23-a045-c0d8c8b8eeb0)

Valemos commented on octoprint_bambu_printer/printer/states/paused_state.py:
```diff
@@ -44,7 +44,7 @@ def _sendPaused(self):
         paused_timer.start()
 
-    def start_new_print(self):
+    def start_resume_print(self):
```
"start_resume_print" does not actually makes sense as state function naming.
I would rather name the gcode handler for starting the print "start_resume" and leave state functions as separate "start" and "resume" functions.
I will submit my version in next commits

Valemos commented:
Fixed the issues you mentioned.

About repeating "Not SD Printing" I think it's sufficient to send this message once when print was finished, since it does not make sense to repeat it unnecessarily, what do you think?
Fixed the print finish. I was thinking to allow user to see, his print finished and deliberately left progress at 100%. But I see that it's confusing, so I will reset the print progress now to make it clear. It also expects a printer to send FINISHED state update message to exit printing job loop, and I'm not sure how to handle 100%. Bambu actually does some work while being at 100%, so it may be too early to close the loop as soon as progress reaches 100

jneilliii commented:
> About repeating "Not SD Printing" I think it's sufficient to send this message once when print was finished, since it does not make sense to repeat it unnecessarily, what do you think?

It should report this when M27 is sent to the virtual printer, requesting status. 

Just tested your latest changes and pause/resume seem to be working fine, I just started a print from the printer, and it seems to have selected the file properly in OctoPrint, but didn't pick up the state change from Operational to Printing. It is reporting printing though, so not sure why the state change isn't getting picked up. Seems like it's finishing before starting. 

```
2024-07-27 21:08:53,651 >>> T:31.00/ 0.00 B:28.00/ 0.00 @:64
2024-07-27 21:08:53,652 >>> ok
2024-07-27 21:08:57,917 >>> File opened: OTF.gcode.3mf  Size: 633732
2024-07-27 21:08:57,920 >>> File selected
2024-07-27 21:08:57,930 >>> Done printing file
2024-07-27 21:08:57,940 >>> Not SD printing
2024-07-27 21:08:58,648 <<< N528 M105*40

2024-07-27 21:08:58,651 >>> T:31.00/ 0.00 B:28.00/ 0.00 @:64
2024-07-27 21:08:58,653 >>> ok
2024-07-27 21:08:58,771 <<< N529 M27*24

2024-07-27 21:08:58,774 >>> ok
2024-07-27 21:08:58,886 >>> File opened: OTF.gcode.3mf  Size: 633732
2024-07-27 21:08:58,889 >>> File selected
2024-07-27 21:08:59,784 <<< N530 M27*16
```

jneilliii commented:
Disconnecting from printer and reconnecting seems to start to pick up the printing state not picked up before, but gets hung up on "Starting print from SD" status in OctoPrint. 

![image](https://github.com/user-attachments/assets/53541bb3-7263-4201-a625-ae88c27af8e1)

This previous to the last changes would have worked to pick up printing state.

Valemos commented:
@jneilliii 
Now print finishing must be detected correctly and state changes mostly come from Printer events (no "zombie" prints this way). Also did some other fixes along with tests.
I cannot get print start command to work on my P1S, maybe MQTT message does not get interpreted correctly (all other commands like pause/resume work fine. Maybe you know what could be the case? I already found this doc page https://github.com/Doridian/OpenBambuAPI/blob/main/cloud-http.md and changed the url multiple times, but printer just ignores that command.

jneilliii commented:
> I already found this doc page [Doridian/OpenBambuAPI@`main`/cloud-http.md](https://github.com/Doridian/OpenBambuAPI/blob/main/cloud-http.md?rgh-link-date=2024-07-28T14%3A33%3A42Z) and changed the url multiple times, but printer just ignores that command.

It's this doc I've used in the past: <https://github.com/Doridian/OpenBambuAPI/blob/main/mqtt.md#printproject_file> The difference between P1S and X1 I think is in how the URL field is entered. The previous logic for the path was [here](https://github.com/jneilliii/OctoPrint-BambuPrinter/blob/ad862d5ebd4b0579d52923c2725ea4c7a7cc137e/octoprint_bambu_printer/virtual.py#L871) in the original code base. This can be impacted if the printer is cloud connected or printing from a file that was in the cache folder. Selected file in OctoPrint could potentially be in cache sub-folder, but the URL isn't pathing it out right and only giving the filename and not full path maybe?

jneilliii commented:
Latest updates seem to be getting into a file listing loop and print started from printer isn't getting picked up. 

[plugin_bambu_printer_serial.log](https://github.com/user-attachments/files/16404641/plugin_bambu_printer_serial.log)

jneilliii commented:
Here's the relevant octoprint.log, where it doesn't see the 3mf and probably starts the file list loop.

[octoprint.log](https://github.com/user-attachments/files/16405276/octoprint.log)

Valemos commented:
@jneilliii ,
For the file loop the issue was when project file is named as "filename.gcode.3mf" the "stem" is defined as filename.gcode instead of expected print job name of "filename". Fixed it and covered with test.
Probably it cannot pick up RUNNING printer state because it was not able to find the file.

jneilliii commented:
> @jneilliii , For the file loop the issue was when project file is named as "filename.gcode.3mf" the "stem" is defined as filename.gcode instead of expected print job name of "filename". Fixed it and covered with test. Probably it cannot pick up RUNNING printer state because it was not able to find the file.

Cool, will test again later tonight when I get back home. This was kind of what I thought was happening because I had that logic in the plugin for testing all the possible variations of subtask_name on detection prior to refreshing the file list. 

https://github.com/jneilliii/OctoPrint-BambuPrinter/blob/ad862d5ebd4b0579d52923c2725ea4c7a7cc137e/octoprint_bambu_printer/virtual.py#L181

jneilliii commented:
file is being detected now and not getting the looping file lists now, but something is still not quite right relative to OctoPrint's printing state when starting from the printer's panel. Still shows as Operational and not detecting start of print.

![image](https://github.com/user-attachments/assets/08c0d005-0f7f-42e3-a6e1-3ba62abb32c8)

```
2024-07-28 22:45:20,432 - octoprint.plugins.bambu_printer.BambuPrinter.serial - DEBUG - >>> File opened: f14 pla rotated.gcode.3mf  Size: 439994
2024-07-28 22:45:20,435 - octoprint.plugins.bambu_printer.BambuPrinter.serial - DEBUG - >>> File selected
```

I actually can't remember if this is a bug from OctoPrint I encountered before relating to filenames with spaces in the name which is where the 8 dot 3 dos naming might come into play. I do remember having the 8.3 filename pointer to the long filename entry in the cached file list. 

https://github.com/jneilliii/OctoPrint-BambuPrinter/blob/ad862d5ebd4b0579d52923c2725ea4c7a7cc137e/octoprint_bambu_printer/virtual.py#L732-L738

jneilliii commented:
Yeah, that may have existed in the current version of the plugin, I'd have to go back and check. The latest changes does pick up on filename without spaces though.

![image](https://github.com/user-attachments/assets/baed046d-2684-420f-a7ec-fc8b1107689a)

Valemos commented:
Maybe you can provide log file for first case? I see on my side that long path with spaces and .gcode.3mf extension gets synced from printer on startup.
Maybe the issue comes when Octoprint is already running and you then start a print.

jneilliii commented:
ok, tried to recreate the issue and it was detected properly this time. not sure if that was a difference in switching from a "FAILED" state from the previous cancelled print or not. Will do some more testing over the next few days.

jneilliii commented:
> ok, tried to recreate the issue and it was detected properly this time. not sure if that was a difference in switching from a "FAILED" state from the previous cancelled print or not. 

That's not it. After the print completed I used print again option on the printer and it was detected the second time as well. Also restarted OctoPrint in the middle of the print and when it came back online once the file list was generated it detect the correct file printing as expected. The previous issue must have been a fluke.
